source/particles/particle_handler.cc

```diff
@@ -426,6 +426,203 @@ namespace Particles
+          Utilities::MPI::some_to_some(triangulation->get_communicator(),
+                                       non_locally_owned_properties);
+
+        // Store all local properties in a single vector. This includes
+        // properties coming from my own mpi process, and properties that
+        // were sent to me in the call above.
+        std::vector<double> local_properties;
+        local_properties.reserve(n_locally_owned_particles() *
+                                 n_properties_per_particle());
+
+        // Compute the association between particle id and start of
+        // property data in the vector containing all local properties
+        std::map<types::particle_index, unsigned int> property_start;
+        for (const auto &it : cpu_to_indices)
+          if (it.first != my_cpu)
+            {
+              unsigned int sequential_index = 0;
+              // Process all properties coming from other mpi processes
+              for (const auto &el : it.second)
+                {
+                  types::particle_index particle_id =
+                    el + starting_points[it.first];
+                  property_start.insert({particle_id, local_properties.size()});
+
+                  local_properties.insert(
+                    local_properties.end(),
+                    locally_owned_properties_from_other_cpus.at(it.first)
+                        .begin() +
+                      sequential_index * n_properties_per_particle(),
+                    locally_owned_properties_from_other_cpus.at(it.first)
+                        .begin() +
+                      (sequential_index + 1) * n_properties_per_particle());
+                  sequential_index++;
+                }
+            }
+          else
+            {
+              // Process all properties that we already own
+              for (const auto &el : it.second)
+                {
+                  types::particle_index particle_id =
+                    el + starting_points[my_cpu];
+                  property_start.insert({particle_id, local_properties.size()});
+
+                  local_properties.insert(local_properties.end(),
+                                          properties.begin() +
+                                            el * n_properties_per_particle(),
+                                          properties.begin() +
+                                            (el + 1) *
+                                              n_properties_per_particle());
+                }
+            }
+        // Actually fill the property pool of each particle.
+        for (auto particle : *this)
+          {
+            particle.set_property_pool(get_property_pool());
+            const auto id = particle.get_id();
+            Assert(property_start.find(id) != property_start.end(),
+                   ExcInternalError());
+            const auto start = property_start[id];
+            particle.set_properties(
```

Comment:
Since the properties are stored by particle anyway, do you need the unified copy of all properties in lines 557-607? Or could you move the logic of where to read the properties from into this loop? This might reduce the number of index variables one has to keep track of.

Reply:
With the some to some doing "the right thing", this would be unified anyway... 

I'd like to try changing the some_to_some in another PR, and then come back to this. It took some time to get this right...